lib/internal/per_context/primordials.js

```diff
@@ -477,6 +477,34 @@ primordials.SafePromiseAll = (promises, mapFn) =>
     SafePromise.all(arrayToSafePromiseIterable(promises, mapFn)).then(a, b)
   );
 
+/**
+ * Should only be used for internal functions, this would produce similar
+ * results as `Promise.all` but without prototype pollution, and the return
+ * value is not a genuine Array but an array-like object.
+ * @param {Promise<any>[]} promises
+ * @returns {Promise<any[]>}
+ */
+primordials.SafePromiseAllReturnArrayLike = async (promises) => {
+  const { length } = promises;
+  const returnVal = { __proto__: null, length };
+  for (let i = 0; i < length; i++) {
+    returnVal[i] = await promises[i];
+  }
+  return returnVal;
+};
+
+/**
+ * Should only be used when we only care about waiting for all the promises to
+ * resolve, not what value they resolve to.
+ * @param {Promise<any>[]} promises
+ * @returns {Promise<void>}
+ */
+primordials.SafePromiseAllReturnVoid = async (promises) => {
+  for (let i = 0; i < promises.length; i++) {
+    await promises[i];
+  }
+};
```

Comment:
this isn't the same thing as Promise.all - this one awaits in serial, but Promise.all rejects whichever is the *first* to reject tick-wise, not index-wise.

Reply:
That's unfortunate, but I don't think there's anything I can do about this.

Reply:
There's no way to prepare a safe iterable, use the for loop to copy promises over to it, and then pass that into `PromiseAll`?

Reply:
Could you maybe adapt a `Promise.all` polyfill to keep the parallelism? For example: https://gist.github.com/Rich-Harris/11010768#file-promise-js-L92-L116

Reply:
After thinking about it a bit more, that's more than just unfortunate, we don't want never-settling promises to block the promise to be rejected. I need to find a workaround.

Reply:
The only Promise.all polyfills I've seen - whether it's that gist, or es6-shim, or core-js - work because they're polyfilling *Promise itself* - i don't know if it's possible to replicate the functionality on native promises without using `PromiseAll` somehow.

I'd love to understand more about the leaks when using PromiseAll - @aduh95, can you elaborate?

Reply:
> I'd love to understand more about the leaks when using PromiseAll - @aduh95, can you elaborate?

See https://tc39.es/ecma262/#sec-promise.all-resolve-element-functions, on step 10, `%Promise.all%` creates an array (that has prototype of `%Array.prototype%`) and pass it to the `promiseCapability.[[Resolve]]` function, which looks up the `then` property on said array. The value of that `then` property is inherited from `%Array.prototype%`, and by extension `%Object.prototype%`.

```js
Array.prototype.then = () => { throw new Error };
Promise.all([]);
```

> The only Promise.all polyfills I've seen - whether it's that gist, or es6-shim, or core-js - work because they're polyfilling _Promise itself_ - i don't know if it's possible to replicate the functionality on native promises without using `PromiseAll` somehow.

PTAL, I think I've managed to came up with something that's close to the original spec (except on the points that would make it vulnerable to prototype pollution).